### ISSUE TITLE
A test suite would help understanding how to use the methods

### ISSUE THREAD

**Issue by stenci** [gileCAD/Gile.AutoCAD.Extension#7]
I often look at the tests to understand how a library is supposed to be used.

It is difficult for me to use these extensions because there are no tests and I don't know how to use them.

For example, I am trying to use the `DBObjectExtension.SetXDictionaryXrecordData` Method. I see that it requires `ResultBuffer data` (or `TypedValue[] values`), but I don't know how to create them. Having a little test that shows how to create them would save me tons of time. This is just an example, the same applies to all other methods in this repository, with the exception of the ones that are very simple and self-explanatory,

I would love to create a PR, but... well, maybe in the future I will be able to do it, right now I will need to work and understand how to use those methods before being able to create the tests :(

**Comment by gileCAD** [comment]
I certainly should have specified it in the README, this library does not exempt from knowing the basics of the AutoCAD .NET API. It is rather aimed at AutoCAD programmers who would like to optimize their development workflow.
For example, the extension method `DBObjectExtension.SetXDictionaryXrecordData` just makes it easier to add or modify an `Xrecord` but its use assumes that you know what an `Xrecord` is and how the data is stored in it, therefore knowing the `ResultBuffer` and `TypedValue` types.

**Comment by stenci** [comment]
> this library does not exempt from knowing the basics of the AutoCAD .NET API

I agree, and it's great as is. The library doesn't need any improvement.

My proposal was about adding some tests so this repository would have one more purpose: it would help learning how to use the library methods and the objects it extends themselves.

The official AutoCAD documentation has very few examples, some are in C++, some of the ones in C# are obsolete, some have syntax errors and don't even compile. The only way to find examples is the forum, but those are often messy.

Perhaps this shouldn't change, perhaps creating a second repository with examples and snippets would be a better solution?

I created this issue when I found [this post](https://forums.autodesk.com/t5/net/class-to-handle-xdictionary-xrecords/m-p/8204596#M60045) and I thought that if that little example had been included in the repository, it would have saved a few precious hours of my life.

**Comment by gileCAD** [comment]
What you would like is not related to the library. 
You give as an example the ResultBuffer and TypedValue types which have no extension methods in this library. 
In fact, you would like a detailed documentation with examples for each type in the API. The task is huge and I don't have the skills to do it.

**Comment by stenci** [comment]
> you would like a detailed documentation with examples for each type in the API

No, I would never ask for that. I would like a few simple tests, as I mentioned earlier.
I didn't ask for documentation and didn't even ask for comments in tests.
In the post I mentioned above you put some description, some code that comes almost straight from this repository and a little test that shows how to use it. 

All I am saying is that after reading that post I can finally use this repository.

I didn't need the description nor the code that is already included in this repository.
All I needed was the test at the bottom. This little command is your code, and in my opinion would add great value to this repository. A test suite that runs without the interactive `GetEntity` might be the right thing to do, but that might also be overkilling and confusing. As is it is already a great help.
```c#
        [CommandMethod("XRECTEST")]
        public static void XrecordTest()
        {
            var doc = Application.DocumentManager.MdiActiveDocument;
            var db = doc.Database;
            var ed = doc.Editor;

            var per = ed.GetEntity("\nSelect object: ");
            if (per.Status == PromptStatus.OK)
            {
                using (var tr = db.TransactionManager.StartTransaction())
                {
                    var ent = (Entity)tr.GetObject(per.ObjectId, OpenMode.ForRead);
                    var data = ent.GetXDictionaryXrecordData("TestXrec");
                    if (data == null)
                    {
                        ed.WriteMessage("\nThe entity does not have a 'TextXrec' Xrecord, it will be added");
                        ent.SetXDictionaryXrecordData("TestXrec", new TypedValue(1, "foo"), new TypedValue(70, 42));
                    }
                    else
                    {
                        foreach (var tv in data.AsArray())
                        {
                            ed.WriteMessage($"\nTypeCode: {tv.TypeCode}, Value: {tv.Value}");
                        }
                    }
                    tr.Commit();
                }
            }
        }
```

I'm sure I'm not the first one that tried to use your library, failed, abandoned, found one of your forum posts, went back to your library and finally got it.

If your repository had this test, you could have answered to that forum post with a link to that test. Faster for you (less typing and no need to create all that code that works without your library) and much much much better for the reader (they would find the same help your post gave, but they would also find other helpful examples).

Yes, I know, providing the code in the forum post is always better than a link to an external resource, but GitHub is an external resource that is not going to disappear any time soon, and the added value provided by your repository with a few test examples would be so high that everybody would be happy about it. I'm pretty sure if you posted a link to the test in the repository instead of writing that long post, that would have been marked as the solution too.

> The task is huge and I don't have the skills to do it

Yeah, right!
You are single handedly running the AutoCad API forum, which is where people go when the official documentation fails. If you don't have the skills to do it (whatever "it" is), no one has!

**Comment by gileCAD** [comment]
I don't quite understand what could have helped you figure out how to use the library in this test example.
In the quoted example, the expression:
`ent.SetXDictionaryXrecordData("TestXrec", new TypedValue(1, "foo"), new TypedValue(70, 42));`
matches exactly the signature of the method:
`public static void SetXDictionaryXrecordData(this DBObject target, string key, params TypedValue[] values)`
Again, if your misunderstanding has to do with `ResultBuffer` or `TypedValue` types, it is not related to this library.

Again, this library does not exempt you from knowing the basics of .NET and the AutoCAD API. If some people don't understand how to use it, it's because they lack those basics.

I do not plan to add test examples to this library, which should remain a pure class library.

**Comment by stenci** [comment]
> I do not plan to add test examples to this library, which should remain a pure class library.

Well this closes the issue 😄

The tests don't need to be part of the class library, they would need to use it. They could be quick and dirty, interactive like the snippet I pasted earlier, and inside the same dll or in another one. Or they could be well designed and do what good tests do, that is starting their own instance of AutoCad and reporting failures if there are any. The latter would be the proper way, but the former would be easier to do.

> I don't quite understand what could have helped you figure out how to use the library in this test example.

I don't know what my misunderstanding came from. All I know is that the test in that post helped me, without it I wouldn't be able to use those methods. Sometimes the signature is enough, sometimes one needs to see a working example, execute it step by step and explore it in the debugger. Much better than looking at a signature.

Here are two use cases for the tests I'm talking about:
1. I wasted hours trying to use those methods. After seeing that little example on that post I figured it out in 5 minutes.
2. You spent some time writing that post. I don't know how much time, but posting a link to the test would have been (1) faster for you and (2) more useful for the reader.

You have shown that you are available and able to help, by publishing this and by helping us all in the forum, but I don't understand (but completely respect) your approach. To me feels like you are saying "hey guys, here is a great library for you to use it, but if you have problems using it you will need to ask in the forum, because I would rather rewrite half the code in the library again and again in the forum so I can help you, then adding a few little tests here".

I'm closing this issue. I will add the tests to my local copy, by copying them from your posts in the forum as I find them.
